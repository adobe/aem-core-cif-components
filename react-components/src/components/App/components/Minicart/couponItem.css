.root {
    display: grid;
    grid-gap: 0 1rem;
    grid-template-columns: 1fr min-content;
    margin: 0 1.5rem;
}
.couponName {
    line-height: 1.5rem;
}
