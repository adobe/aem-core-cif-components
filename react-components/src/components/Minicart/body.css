.root {
    display: grid;
    grid-template-rows: min-content min-content 1fr;
    overflow: auto;
}
