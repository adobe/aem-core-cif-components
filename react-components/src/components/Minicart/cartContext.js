/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/

import React, { createContext, useContext, useReducer } from 'react';
import { object, func } from 'prop-types';
import { useCookieValue } from '../../utils/hooks';

export const initialState = {
    isOpen: false,
    isRegistered: false,
    isEditing: false,
    isLoading: false,
    editItem: {},
    cartId: null,
    cart: null,
    errorMessage: null,
    couponError: null
};

export const reducerFactory = setCartCookie => {
    return (state, action) => {
        switch (action.type) {
            case 'close':
                return {
                    ...state,
                    isOpen: false
                };
            case 'open':
                return {
                    ...state,
                    isOpen: true
                };
            case 'beginLoading':
                return {
                    ...state,
                    isLoading: true
                };
            case 'endLoading':
                return {
                    ...state,
                    isLoading: false
                };
            case 'beginEditing':
                return {
                    ...state,
                    isEditing: true,
                    editItem: action.item
                };
            case 'endEditing':
                return {
                    ...state,
                    isEditing: false,
                    editItem: {}
                };
            case 'cartId':
                return {
                    ...state,
                    cartId: action.cartId
                };
            case 'reset':
                setCartCookie('', 0);
                return {
                    ...state,
                    cartId: null,
                    errorMessage: null,
                    couponError: null,
                    cart: null
                };
            case 'cart':
                return {
                    ...state,
                    cart: action.cart,
                    isLoading: false,
                    couponError: !action.cart.applied_coupon ? state.couponError : null
                };
            case 'register': {
                return {
                    ...state,
                    isRegistered: true
                };
            }
            case 'error':
                console.error(action.error);
                return {
                    ...state,
                    errorMessage: action.error
                };
            case 'discardError':
                return {
                    ...state,
                    errorMessage: null
                };
            case 'couponError':
                return {
                    ...state,
                    couponError: action.error
                };

            default:
                return state;
        }
    };
};

export const CartContext = createContext();

export const CartProvider = props => {
    const factory = props.reducerFactory || reducerFactory;
    const state = props.initialState || initialState;

    const [, setCartCookie] = useCookieValue('cif.cart');
    const contextValue = useReducer(factory(setCartCookie), state);

    return <CartContext.Provider value={contextValue}>{props.children}</CartContext.Provider>;
};

CartProvider.propTypes = {
    reducerFactory: func,
    initialState: object
};

export const useCartState = () => useContext(CartContext);
