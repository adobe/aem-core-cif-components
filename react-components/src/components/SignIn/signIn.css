.root {
    display: grid;
    gap: 1.5rem;
    justify-items: stretch;
    padding: 1rem 1.5rem;
}

.forgotPassword {
    display: block;
    text-decoration: underline;
    color: rgb(var(--venia-error));
    font-size: 0.8rem;
}

.form {
    display: grid;
    row-gap: 1rem;
}

.formTitle {
    padding: 1rem 0 0.5rem;
}

.modal {
    visibility: hidden;
    height: 100%;
    width: 100%;
    background-color: rgb(var(--venia-grey));
    text-align: center;
    position: absolute;
    bottom: 0;
}

.modal_active {
    visibility: visible;
    composes: modal;
    opacity: 0.9;
}

.createAccountButton,
.forgotPasswordButton,
.signInButton {
    align-items: center;
    display: flex;
    text-align: center;
}

.createAccountButton,
.signInButton {
    justify-content: center;
}

.createAccountButton button,
.signInButton button {
    min-width: 10rem;
}

.forgotPasswordButton {
    justify-content: flex-start;
}

.signInButton {
    margin-top: 1rem;
}

.signInError {
    background-color: rgb(var(--venia-grey));
    border-radius: 4px;
    color: rgb(var(--venia-error));
    font-size: 0.875rem;
    line-height: 1.25rem;
    padding: 1rem;
}

.signInError:empty {
    display: none;
}
