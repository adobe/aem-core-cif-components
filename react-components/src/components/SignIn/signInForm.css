.form {
    display: grid;
    row-gap: 1rem;
}

.signInError {
    background-color: rgb(var(--venia-grey));
    border-radius: 4px;
    color: rgb(var(--venia-error));
    font-size: 0.875rem;
    line-height: 1.25rem;
    padding: 1rem;
}

.signInError:empty {
    display: none;
}

.signInButton {
    align-items: center;
    display: flex;
    justify-content: center;
    text-align: center;
    margin-top: 1rem;
}
