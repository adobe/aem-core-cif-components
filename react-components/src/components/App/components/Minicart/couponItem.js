/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { useCartState } from './cartContext';
import Kebab from './kebab';
import Section from './section';

import classes from './couponItem.css';

const CouponItem = () => {
    const [{ cart, removeCoupon }] = useCartState();

    const appliedCoupon = cart.applied_coupon ? cart.applied_coupon.code : null;

    return (
        <div className={classes.root}>
            <div className={classes.couponName}>
                Coupon <strong>{appliedCoupon}</strong> applied.
            </div>
            <Kebab>
                <Section text="Remove coupon" onClick={removeCoupon} icon="Trash" />
            </Kebab>
        </div>
    );
};

export default CouponItem;
