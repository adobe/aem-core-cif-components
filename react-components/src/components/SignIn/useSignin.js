/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import { useState } from 'react';
import { useUserContext } from '../../context/UserContext';
import { useMutation } from '@apollo/client';
import { useCartState } from '../Minicart/cartContext';
import { useAwaitQuery, useCookieValue } from '../../utils/hooks';
import { mergeCarts } from '../../actions/cart';

import MUTATION_MERGE_CARTS from '../../queries/mutation_merge_carts.graphql';
import QUERY_CUSTOMER_CART from '../../queries/query_customer_cart.graphql';
import MUTATION_GENERATE_TOKEN from '../../queries/mutation_generate_token.graphql';
import QUERY_CART_DETAILS from '../../queries/query_cart_details.graphql';

export const useSignin = props => {
    const { showMyAccount } = props;
    const [{ cartId }, cartDispatch] = useCartState();
    const [userState, { setToken, getUserDetails, setCustomerCart, setError }] = useUserContext();
    const [inProgress, setInProgress] = useState(false);

    const [, setCartCookie] = useCookieValue('cif.cart');

    const [mergeCartsMutation] = useMutation(MUTATION_MERGE_CARTS);
    const fetchCustomerCart = useAwaitQuery(QUERY_CUSTOMER_CART);
    const cartDetailsQuery = useAwaitQuery(QUERY_CART_DETAILS);
    const [generateCustomerToken] = useMutation(MUTATION_GENERATE_TOKEN);

    let errorMessage = '';
    if (userState.signInError && userState.signInError.length > 0) {
        errorMessage = userState.signInError;
    }

    const handleSubmit = async ({ email, password }) => {
        setInProgress(true);
        try {
            // 1. generate the customer token
            const { data } = await generateCustomerToken({ variables: { email, password } });
            const token = data.generateCustomerToken.token;
            setToken(token);

            // 2. get the user details
            await getUserDetails();

            const { data: customerCartData } = await fetchCustomerCart({
                fetchPolicy: 'network-only'
            });
            const customerCartId = customerCartData.customerCart.id;

            // 3. merge the shopping cart if necessary
            let mergedCartId;

            if (cartId) {
                mergedCartId = await mergeCarts({
                    mergeCartsMutation,
                    cartDetailsQuery,
                    cartId,
                    customerCartId,
                    dispatch: cartDispatch
                });
            } else {
                mergedCartId = customerCartId;
            }
            //4. set the cart id in the cookie
            setCartCookie(mergedCartId);
            setCustomerCart(mergedCartId);

            //5. show my account view in account dropdown or navigation side panel after sign in
            showMyAccount();
        } catch (e) {
            setError(e);
            setInProgress(false);
        }
    };

    return {
        inProgress,
        handleSubmit,
        errorMessage,
        isSignedIn: userState.isSignedIn
    };
};
