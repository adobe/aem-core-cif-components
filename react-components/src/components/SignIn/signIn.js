/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { Form } from 'informed';
import { func } from 'prop-types';
import { useTranslation } from 'react-i18next';

import { isRequired } from '../../utils/formValidators';
import Button from '../Button';
import Field from '../Field';
import TextInput from '../TextInput';

import classes from './signIn.css';
import { useSignin } from './useSignin';
import LoadingIndicator from '../LoadingIndicator';

const SignIn = props => {
    const { showMyAccount, showForgotPassword, showCreateAccount } = props;
    const { errorMessage, handleSubmit, inProgress } = useSignin({ showMyAccount });
    const [t] = useTranslation('account');

    if (inProgress) {
        return <LoadingIndicator>{t('account:signing-in', 'Signing In')}</LoadingIndicator>;
    }

    return (
        <div className={classes.root}>
            <Form onSubmit={handleSubmit} className={classes.form}>
                <div className={classes.formTitle}>{t('account:sign-in-form-title', 'Sign-in to Your Account')}</div>
                <Field label={t('account:email', 'Email address')} required={true}>
                    <TextInput autoComplete="email" field="email" validate={isRequired} aria-label="email" />
                </Field>
                <Field label={t('account:password', 'Password')} required={true}>
                    <TextInput
                        autoComplete="current-password"
                        field="password"
                        type="password"
                        validate={isRequired}
                        aria-label="password"
                    />
                </Field>
                <div className={classes.forgotPasswordButton}>
                    <Button priority="low" type="button" onClick={showForgotPassword}>
                        {t('account:forgot-password', 'Forgot Password?')}
                    </Button>
                </div>
                <div className={classes.signInError}>{errorMessage}</div>
                <div className={classes.signInButton}>
                    <Button priority="high" type="submit" aria-label="submit">
                        {t('account:sign-in', 'Sign In')}
                    </Button>
                </div>
                <div className={classes.createAccountButton}>
                    <Button priority="normal" type="button" onClick={showCreateAccount}>
                        {t('account:create-an-account', 'Create an Account')}
                    </Button>
                </div>
            </Form>
        </div>
    );
};

SignIn.propTypes = {
    showMyAccount: func.isRequired,
    showForgotPassword: func.isRequired,
    showCreateAccount: func.isRequired
};

export default SignIn;
