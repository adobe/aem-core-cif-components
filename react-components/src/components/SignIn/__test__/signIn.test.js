/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { fireEvent, wait } from '@testing-library/react';
import { render } from 'test-utils';

import { useUserContext } from '../../../context/UserContext';
import { CartProvider } from '../../Minicart/cartContext';

import SignIn from '../signIn';

describe('<SignIn>', () => {
    beforeEach(() => {
        Object.defineProperty(window.document, 'cookie', {
            writable: true,
            value: 'cif.cart=guest123'
        });
    });

    it('renders the component', () => {
        const { asFragment } = render(
            <CartProvider initialState={{ cartId: null }} reducerFactory={() => state => state}>
                <SignIn showMyAccount={jest.fn()} showCreateAccount={jest.fn()} showForgotPassword={jest.fn()} />
            </CartProvider>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it('switch the view then the "Sign In" is successful', async () => {
        // To simulate an almost real use case of the sign in component we create a wrapper around it
        // which displays a "success" message when the user is signed in
        const SignInWrapper = () => {
            const [{ isSignedIn, cartId }] = useUserContext();

            let content;
            if (isSignedIn && cartId) {
                content = <div data-testid="success">Done</div>;
            } else {
                content = (
                    <SignIn showMyAccount={jest.fn()} showCreateAccount={jest.fn()} showForgotPassword={jest.fn()} />
                );
            }

            return <div>{content}</div>;
        };

        const { getByTestId, getByLabelText } = render(
            <CartProvider initialState={{ cartId: null }} reducerFactory={() => state => state}>
                <SignInWrapper />
            </CartProvider>
        );

        fireEvent.change(getByLabelText(/email/i), { target: { value: 'chuck@example.com' } });
        fireEvent.change(getByLabelText(/password/i), { target: { value: 'norris' } });
        fireEvent.click(getByLabelText('submit'));
        await wait(() => {
            expect(getByTestId('success').textContent).not.toBeUndefined();
        });
    });

    it('shows an error when the sign in is not successful', async () => {
        const { getByText, getByLabelText } = render(
            <CartProvider initialState={{ cartId: null }} reducerFactory={() => state => state}>
                <SignIn showMyAccount={jest.fn()} showForgotPassword={jest.fn()} showCreateAccount={jest.fn()} />
            </CartProvider>
        );

        fireEvent.change(getByLabelText(/email/i), { target: { value: 'chuck@example.com' } });
        fireEvent.change(getByLabelText(/password/i), { target: { value: 'wrongpassword' } });
        fireEvent.click(getByLabelText('submit'));

        await wait(() => {
            expect(getByText('Error')).not.toBeUndefined();
        });
    });
});
