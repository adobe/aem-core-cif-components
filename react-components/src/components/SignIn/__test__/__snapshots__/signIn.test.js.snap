// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignIn> renders the component 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <form
      class="form"
    >
      <div
        class="root"
      >
        <span
          class="label"
        >
          <span
            class="requiredSymbol"
          />
          Email
        </span>
        <span
          class="root"
        >
          <span
            class="input"
          >
            <input
              aria-label="email"
              autocomplete="email"
              class="input"
              name="email"
              value=""
            />
          </span>
          <span
            class="before"
          />
          <span
            class="after"
          />
        </span>
        <p
          class="root"
        />
      </div>
      <div
        class="root"
      >
        <span
          class="label"
        >
          <span
            class="requiredSymbol"
          />
          Password
        </span>
        <span
          class="root"
        >
          <span
            class="input"
          >
            <input
              aria-label="password"
              autocomplete="current-password"
              class="input"
              name="password"
              type="password"
              value=""
            />
          </span>
          <span
            class="before"
          />
          <span
            class="after"
          />
        </span>
        <p
          class="root"
        />
      </div>
      <div
        class="signInError"
      />
      <div
        class="signInButton"
      >
        <button
          aria-label="submit"
          class="root_highPriority"
          type="submit"
        >
          <span
            class="content"
          >
            Sign In
          </span>
        </button>
      </div>
      <div
        class="forgotPasswordButton"
      >
        <button
          class="root_lowPriority"
          type="button"
        >
          <span
            class="content"
          >
            Forgot Password?
          </span>
        </button>
      </div>
      <div
        class="signInDivider"
      />
      <div
        class="createAccountButton"
      >
        <button
          class="root_normalPriority"
          type="button"
        >
          <span
            class="content"
          >
            Create an Account
          </span>
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
